pallets/order-book/src/lib.rs

```diff
@@ -571,87 +590,86 @@ pub mod pallet {
+			Self::fulfill_order_with_amount(order, amount_out, account_id)
 		}
 	}
 
 	impl<T: Config> Pallet<T> {
 		pub fn fulfill_order_with_amount(
 			order: OrderOf<T>,
-			buy_amount: T::Balance,
+			amount_out: T::Balance,
 			account_id: T::AccountId,
 		) -> DispatchResult {
 			ensure!(
-				buy_amount >= order.min_fulfillment_amount,
+				amount_out >= order.min_fulfillment_amount,
 				Error::<T>::InsufficientOrderSize,
 			);
 
-			ensure!(
-				T::TradeableAsset::can_hold(order.asset_in_id, &(), &account_id, buy_amount),
-				Error::<T>::InsufficientAssetFunds,
-			);
+			let price = match order.price {
+				OrderPrice::Market => Self::market_price(order.asset_out_id, order.asset_in_id)?,
+				OrderPrice::Custom(price) => price,
+			};
+
+			let amount_in =
+				Self::convert_with_ratio(order.asset_out_id, order.asset_in_id, price, amount_out)?;
 
-			let sell_amount = Self::convert_with_ratio(
-				order.asset_in_id,
-				order.asset_out_id,
-				order.max_sell_rate,
-				buy_amount,
-			)?;
 			let remaining_buy_amount = order
-				.buy_amount
-				.checked_sub(&buy_amount)
+				.amount_out
+				.checked_sub(&amount_out)
 				.ok_or(Error::<T>::BuyAmountTooLarge)?;
-			let partial_fulfillment = !remaining_buy_amount.is_zero();
 
+			let partial_fulfillment = !remaining_buy_amount.is_zero();
 			if partial_fulfillment {
-				Self::update_order_with_fulfillment(
-					order.placing_account.clone(),
-					order.order_id,
-					remaining_buy_amount,
-					order.max_sell_rate,
-					remaining_buy_amount.min(order.min_fulfillment_amount),
-				)?;
+				let mut updated_order = order.clone();
+				updated_order.amount_out = remaining_buy_amount;
+				updated_order.amount_in = order.amount_in.ensure_add(amount_in)?;
+				updated_order.min_fulfillment_amount =
+					remaining_buy_amount.min(order.min_fulfillment_amount);
+
+				<Orders<T>>::insert(updated_order.order_id, updated_order.clone());
+				<UserOrders<T>>::insert(&account_id, updated_order.order_id, updated_order);
 			} else {
-				T::TradeableAsset::release(
-					order.asset_out_id,
-					&(),
-					&order.placing_account,
-					sell_amount,
-					Precision::Exact,
-				)?;
-
 				Self::remove_order(order.order_id)?;
 			}
 
-			T::TradeableAsset::transfer(
-				order.asset_in_id,
-				&account_id,
+			T::TradeableAsset::release(
+				order.asset_out_id,
+				&(),
 				&order.placing_account,
-				buy_amount,
-				Preservation::Expendable,
+				amount_out,
+				Precision::Exact,
 			)?;
 			T::TradeableAsset::transfer(
 				order.asset_out_id,
 				&order.placing_account,
 				&account_id,
-				sell_amount,
+				amount_out,
+				Preservation::Expendable,
+			)?;
+			T::TradeableAsset::transfer(
+				order.asset_in_id,
+				&account_id,
+				&order.placing_account,
+				amount_in,
 				Preservation::Expendable,
 			)?;
 
 			T::FulfilledOrderHook::notify_status_change(
 				order.order_id,
 				Swap {
-					amount: buy_amount,
+					amount: todo!(),
```

Comment:
@wischli I'm not sure what I should put here after these changes.

Before, it was expected to be the `buy_amount`, but now we have no way to know it for market prices. Should I change it to `amount_out`? (which is `sell_amount`?)

Reply:
It should be `amount_in` because this value is needed to update investment states with pending swaps. The outgoing amount is already known.  The `amount_out` part does not help too much because then Foreign Investments would have to determine the `amount_in` from the `amount_out`.

Reply:
Previously, the `buy_amount` or `amount_in` was known. But now with a variable market price, we do not know `amount_in`. It meant that you still need to fulfill `amount_in`. Now `amount_in` corresponds to the current already fulfilled amount. It's different to what it meant before.

Reply:
But the `amount_in` (resp. `buy_amount`) is known here because we have fulfilled an order?! The purpose of the hook is to inform FI that a (partial) swap was fulfilled for which amount.

Reply:
> Now `amount_in` corresponds to the current already fulfilled amount. It's different to what it meant before.

I disagree. In L625 we determine the `amount_in` from the `amount_out` which is the fulfilled amount at this point in time (which is reflected by determining the remaining order amount by reducing its previous value by `amount_out` in L628) : 
```
let amount_in =
				Self::convert_with_ratio(order.asset_out_id, order.asset_in_id, price, amount_out)?;
```

So `amount_in` is exactly the fulfilled amount of incoming asset which should be reported to the consumer of this hook (FI) to inform that a (partial) swap was fulfilled for that incoming amount.

Reply:
You're right! I was seeing this wrong. I thought the previous value was `order.buy_amount,`, but it's the `buy_amount` used to fulfill. So translated now is just `amount_in` as you say.

 Thanks a lot for guiding me to the correct reasoning! 🙌🏻